### PR TITLE
[Merged by Bors] - feat: port Order.LiminfLimsup

### PR DESCRIPTION
**Pull request by LukasMias**
Most notable parts of this file:
- Needed to translate `isBoundedDefault` macro to lean4, very easy
- Had to replace `e` with `FunLike.coe e` in a theorem statement to prevent strange coercion behaviour for Equivs.

---

- [x] depends on: #2054 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2054
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2054~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Order/LiminfLimsup.lean [hunk 1]
```diff
@@ -0,0 +1,1242 @@
...
+  isBoundedUnder_le_sup.trans <| and_congr Iff.rfl isBoundedUnder_le_neg
+#align filter.is_bounded_under_le_abs Filter.isBoundedUnder_le_abs
+
+/-- Filters are automatically bounded or cobounded in complete lattices. To use the same statements
+in complete and conditionally complete lattices but let automation fill automatically the
+boundedness proofs in complete lattices, we use the tactic `isBounded_default` in the statements,
+in the form `(hf : f.IsBounded (≥) . isBoundedDefault)`. -/
+
+macro "isBoundedDefault ": tactic =>
+  `(tactic| ( first
```
Comment (jcommelin) [inline Mathlib/Order/LiminfLimsup.lean:320]:
```suggestion
  `(tactic| (first
```

File: Mathlib/Order/LiminfLimsup.lean [hunk 2]
```diff
@@ -0,0 +1,1242 @@
...
+#align filter.limsup_eq_Inf_Sup Filter.limsup_eq_infₛ_supₛ
+
+theorem liminf_eq_supₛ_infₛ {ι R : Type _} (F : Filter ι) [CompleteLattice R] (a : ι → R) :
+    liminf a F = supₛ ((fun I => infₛ (a '' I)) '' F.sets) :=
+  @Filter.limsup_eq_infₛ_supₛ ι (OrderDual R) _ _ a
+set_option linter.uppercaseLean3 false in
+#align filter.liminf_eq_Sup_Inf Filter.liminf_eq_supₛ_infₛ
+
+-- Porting note: simp_nf linter, Left-hand side does not simplify when using simp on itself.
+-- @[simp]
```
Comment (jcommelin) [inline Mathlib/Order/LiminfLimsup.lean:805]: Did you test whether the linter is actually right? If it's claim is false, please add a nolint (and leave a porting note that mentions why).

File: Mathlib/Order/LiminfLimsup.lean [hunk 3]
```diff
@@ -0,0 +1,1242 @@
...
+theorem liminf_nat_add (f : ℕ → α) (k : ℕ) : liminf (fun i => f (i + k)) atTop = liminf f atTop :=
+  by
+  simp_rw [liminf_eq_supᵢ_infᵢ_of_nat]
+  exact supᵢ_infᵢ_ge_nat_add f k
+#align filter.liminf_nat_add Filter.liminf_nat_add
+
+-- Porting note: simp_nf linter, Left-hand side does not simplify when using simp on itself.
+-- @[simp]
```
Comment (jcommelin) [inline Mathlib/Order/LiminfLimsup.lean:813]: Idem

File: Mathlib/Order/LiminfLimsup.lean [hunk 4]
```diff
@@ -0,0 +1,1242 @@
...
+theorem blimsup_and_le_inf : (blimsup u f fun x => p x ∧ q x) ≤ blimsup u f p ⊓ blimsup u f q :=
+  le_inf (blimsup_mono <| by tauto) (blimsup_mono <| by tauto)
+#align filter.blimsup_and_le_inf Filter.blimsup_and_le_inf
+
+@[simp]
+theorem bliminf_sup_le_inf_aux :
+  (blimsup u f fun x => p x ∧ q x) ≤ blimsup u f p
+    ∧ (blimsup u f fun x => p x ∧ q x) ≤ blimsup u f q := by
+  rw [←le_inf_iff]
+  exact blimsup_and_le_inf
```
Comment (jcommelin) [inline Mathlib/Order/LiminfLimsup.lean:901]:
```suggestion
-- Porting note: new lemma
@[simp]
theorem bliminf_sup_le_inf_aux :
  (blimsup u f fun x => p x ∧ q x) ≤ blimsup u f p
    ∧ (blimsup u f fun x => p x ∧ q x) ≤ blimsup u f q := by
  rw [←le_inf_iff]
  exact blimsup_and_le_inf
```
I think you can make simp even more powerful by splitting this into two lemmas: one for the left hand side and one for the right hand side.

File: Mathlib/Order/LiminfLimsup.lean [hunk 5]
```diff
@@ -0,0 +1,1242 @@
...
+
+-- @[simp] -- Porting note: simp_nf linter, lhs simplifies, added _aux simp version below
+theorem bliminf_sup_le_and : bliminf u f p ⊔ bliminf u f q ≤ bliminf u f fun x => p x ∧ q x :=
+  blimsup_and_le_inf (α := αᵒᵈ)
+#align filter.bliminf_sup_le_and Filter.bliminf_sup_le_and
+
+@[simp]
+theorem bliminf_sup_le_and_aux :
```
Comment (jcommelin) [inline Mathlib/Order/LiminfLimsup.lean:909]: Idem.

File: Mathlib/Order/LiminfLimsup.lean [hunk 6]
```diff
@@ -0,0 +1,1242 @@
...
+  exact blimsup_sup_le_or
+
+/-- See also `Filter.bliminf_or_eq_inf`. -/
+-- @[simp] -- Porting note: simp_nf linter, lhs simplifies, added _aux simp version below
+theorem bliminf_or_le_inf : (bliminf u f fun x => p x ∨ q x) ≤ bliminf u f p ⊓ bliminf u f q :=
+  blimsup_sup_le_or (α := αᵒᵈ)
+#align filter.bliminf_or_le_inf Filter.bliminf_or_le_inf
+
+@[simp]
+theorem bliminf_or_le_inf_aux :
```
Comment (jcommelin) [inline Mathlib/Order/LiminfLimsup.lean:935]: Idem.

File: Mathlib/Order/LiminfLimsup.lean [hunk 7]
```diff
@@ -0,0 +1,1242 @@
...
+  (bliminf u f fun x => p x ∨ q x) ≤ bliminf u f p
+    ∧ (bliminf u f fun x => p x ∨ q x) ≤ bliminf u f q := by
+  rw [←le_inf_iff]
+  exact bliminf_or_le_inf
+
+/- Porting note: Replaced `e` with `FunLike.coe e` to override the strange
+ coercion to `↑(RelIso.toRelEmbedding e).toEmbedding`.-/
+theorem OrderIso.apply_blimsup [CompleteLattice γ] (e : α ≃o γ) :
+    FunLike.coe e (blimsup u f p) = blimsup ((FunLike.coe e) ∘ u) f p := by
```
Comment (jcommelin) [inline Mathlib/Order/LiminfLimsup.lean:944]: Does it work to write `(e : α -> γ)` instead of the `FunLike.coe`? That's a bit more readable.

Comment (LukasMias) [reply Mathlib/Order/LiminfLimsup.lean:944]: Thanks for all the very good comments! This one is the only one that doesn't work, `(e : α -> γ)` is automatically coerced to the unpleasant `↑(RelIso.toRelEmbedding e).toEmbedding` expression. This problem is also discussed in [this Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/CoeFun), and due to an undesirable CoeFun instance.

File: Mathlib/Order/LiminfLimsup.lean (hunk 2 above)
Comment (LukasMias) [reply Mathlib/Order/LiminfLimsup.lean:805]: I had too much trust in the linter, it is indeed incorrect here. Thanks!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**